// mini-css-extract-plugin/dist/utils.js
"use strict";

const NativeModule = require("module");
const path = require("path");

/** @typedef {import("webpack").Compilation} Compilation */
/** @typedef {import("webpack").Module} Module */

// eslint-disable-next-line jsdoc/no-restricted-syntax
/** @typedef {import("webpack").LoaderContext<any>} LoaderContext */

/**
 * @returns {boolean} always returns true
 */
function trueFn() {
  return true;
}

/**
 * @param {Compilation} compilation compilation
 * @param {string | number} id module id
 * @returns {null | Module} the found module
 */
function findModuleById(compilation, id) {
  const {
    modules,
    chunkGraph
  } = compilation;
  for (const module of modules) {
    const moduleId = typeof chunkGraph !== "undefined" ? chunkGraph.getModuleId(module) : module.id;
    if (moduleId === id) {
      return module;
    }
  }
  return null;
}

// eslint-disable-next-line jsdoc/no-restricted-syntax
/**
 * @param {LoaderContext} loaderContext loader context
 * @param {string | Buffer} code code
 * @param {string} filename filename
 * @returns {Record<string, any>} exports of a module
 */
function evalModuleCode(loaderContext, code, filename) {
  // @ts-expect-error
  const module = new NativeModule(filename, loaderContext);
  // @ts-expect-error
  module.paths = NativeModule._nodeModulePaths(loaderContext.context);
  module.filename = filename;
  // @ts-expect-error
  module._compile(code, filename);
  return module.exports;
}

/**
 * @param {string} a a
 * @param {string} b b
 * @returns {0 | 1 | -1} result of comparing
 */
function compareIds(a, b) {
  if (typeof a !== typeof b) {
    return typeof a < typeof b ? -1 : 1;
  }
  if (a < b) {
    return -1;
  }
  if (a > b) {
    return 1;
  }
  return 0;
}

/**
 * @param {Module} a a
 * @param {Module} b b
 * @returns {0 | 1 | -1} result of comparing
 */
function compareModulesByIdentifier(a, b) {
  return compareIds(a.identifier(), b.identifier());
}
const MODULE_TYPE = "css/mini-extract";
const AUTO_PUBLIC_PATH = "__mini_css_extract_plugin_public_path_auto__";
const ABSOLUTE_PUBLIC_PATH = "webpack:///mini-css-extract-plugin/";
const BASE_URI = "webpack://";
const SINGLE_DOT_PATH_SEGMENT = "__mini_css_extract_plugin_single_dot_path_segment__";

/**
 * @param {string} str path
 * @returns {boolean} true when path is absolute, otherwise false
 */
function isAbsolutePath(str) {
  return path.posix.isAbsolute(str) || path.win32.isAbsolute(str);
}
const RELATIVE_PATH_REGEXP = /^\.\.?[/\\]/;

/**
 * @param {string} str string
 * @returns {boolean} true when path is relative, otherwise false
 */
function isRelativePath(str) {
  return RELATIVE_PATH_REGEXP.test(str);
}

// TODO simplify for the next major release
/**
 * @param {LoaderContext} loaderContext the loader context
 * @param {string} request a request
 * @returns {string} a stringified request
 */
function stringifyRequest(loaderContext, request) {
  if (typeof loaderContext.utils !== "undefined" && typeof loaderContext.utils.contextify === "function") {
    return JSON.stringify(loaderContext.utils.contextify(loaderContext.context || loaderContext.rootContext, request));
  }
  const splitted = request.split("!");
  const {
    context
  } = loaderContext;
  return JSON.stringify(splitted.map(part => {
    // First, separate singlePath from query, because the query might contain paths again
    const splittedPart = part.match(/^(.*?)(\?.*)/);
    const query = splittedPart ? splittedPart[2] : "";
    let singlePath = splittedPart ? splittedPart[1] : part;
    if (isAbsolutePath(singlePath) && context) {
      singlePath = path.relative(context, singlePath);
      if (isAbsolutePath(singlePath)) {
        // If singlePath still matches an absolute path, singlePath was on a different drive than context.
        // In this case, we leave the path platform-specific without replacing any separators.
        // @see https://github.com/webpack/loader-utils/pull/14
        return singlePath + query;
      }
      if (isRelativePath(singlePath) === false) {
        // Ensure that the relative path starts at least with ./ otherwise it would be a request into the modules directory (like node_modules).
        singlePath = `./${singlePath}`;
      }
    }
    return singlePath.replace(/\\/g, "/") + query;
  }).join("!"));
}

/**
 * @param {string} filename filename
 * @param {string} outputPath output path
 * @param {boolean} enforceRelative true when need to enforce relative path, otherwise false
 * @returns {string} undo path
 */
function getUndoPath(filename, outputPath, enforceRelative) {
  let depth = -1;
  let append = "";
  outputPath = outputPath.replace(/[\\/]$/, "");
  for (const part of filename.split(/[/\\]+/)) {
    if (part === "..") {
      if (depth > -1) {
        depth--;
      } else {
        const i = outputPath.lastIndexOf("/");
        const j = outputPath.lastIndexOf("\\");
        const pos = i < 0 ? j : j < 0 ? i : Math.max(i, j);
        if (pos < 0) {
          return `${outputPath}/`;
        }
        append = `${outputPath.slice(pos + 1)}/${append}`;
        outputPath = outputPath.slice(0, pos);
      }
    } else if (part !== ".") {
      depth++;
    }
  }
  return depth > 0 ? `${"../".repeat(depth)}${append}` : enforceRelative ? `./${append}` : append;
}

// eslint-disable-next-line jsdoc/no-restricted-syntax
/**
 * @param {string | Function} value local
 * @returns {string} stringified local
 */
function stringifyLocal(value) {
  return typeof value === "function" ? value.toString() : JSON.stringify(value);
}

/**
 * @param {string} str string
 * @returns {string} string
 */
const toSimpleString = str => {
  // eslint-disable-next-line no-implicit-coercion
  if (`${+str}` === str) {
    return str;
  }
  return JSON.stringify(str);
};

/**
 * @param {string} str string
 * @returns {string} quoted meta
 */
const quoteMeta = str => str.replace(/[-[\]\\/{}()*+?.^$|]/g, "\\$&");

/**
 * @param {Array<string>} items items
 * @returns {string} common prefix
 */
const getCommonPrefix = items => {
  let [prefix] = items;
  for (let i = 1; i < items.length; i++) {
    const item = items[i];
    for (let prefixIndex = 0; prefixIndex < prefix.length; prefixIndex++) {
      if (item[prefixIndex] !== prefix[prefixIndex]) {
        prefix = prefix.slice(0, prefixIndex);
        break;
      }
    }
  }
  return prefix;
};

/**
 * @param {Array<string>} items items
 * @returns {string} common suffix
 */
const getCommonSuffix = items => {
  let [suffix] = items;
  for (let i = 1; i < items.length; i++) {
    const item = items[i];
    for (let itemIndex = item.length - 1, suffixIndex = suffix.length - 1; suffixInde
